### DISCUSSION TITLE
Access length of spline for displaying well logs and well tops

### DISCUSSION THREAD

**Discussion by AlexanderJuestel** [pyvista/pyvista#3123] [Ideas]
Hello, 

For structural geological modeling and visualizing well log data, I would like to be able to index the spline by its length. 

Let's take the spline from the docs and assume it is a well/borehole that was drilled. In geosciences, we now map the boundaries of different layers (displayed as red dots) according to the distance from the top of the borehole (the yellow part), here let's say 50 m, 100 m, and 300 m. I would now like to be able to index the spline to access the XYZ coordinates of the point where the spline is at a length of the aforementioned values as these would be the true/real-world XYZ coordinates that we would need for geomodeling.

![image](https://user-images.githubusercontent.com/45469915/182553727-bce13ada-38a4-4f13-85c8-b2ffd87cf660.png)

Adding to this, I would like to connect the depth of the well/the length of the spline with a geophysical measurement (the radius of the tube is equal to the measurement values) which is shown in the DataFrame and the image below but it is falsely mapped along the spline. Rather than placing the values of 'GR_EDTC' at the correct measured depth along the spline ('MD'), it was placed equally along the 10243 elements of the spline. However, most of the times, not the entire well was logged so it would be necessary to access the depth of the spline once again. 

I hope my explanations were sufficient to understand the issue. Let me know if you have any further questions and I am happy for any tips and tricks. 

Cheers 
Alex

![image](https://user-images.githubusercontent.com/45469915/182554733-07a06cb7-bbb5-4ff1-bb60-1ec68bb75f8a.png)

![image](https://user-images.githubusercontent.com/45469915/182554386-e9154333-e2d2-4fbb-97d5-0969fb3d2940.png)

**Comment by akaszynski** [comment]
If you have a pre-calculated arc-length, it's not too hard getting the nearest point matching that location:

![tmp](https://user-images.githubusercontent.com/11981631/182759109-3380a94c-8f7f-4db1-9020-5803ca3c17fa.png)


```py
import numpy as np
import pyvista as pv
pv.set_plot_theme('document')

# Make the xyz points
theta = np.linspace(-1 * np.pi, 1 * np.pi, 100)
z = np.linspace(2, -2, 100)
r = z**2 + 1
x = r * np.sin(theta)
y = r * np.cos(theta)
points = np.column_stack((x, y, z))

# Create a spline. This automatically computes arc_length, which is the distance along the line.
spline = pv.Spline(points, 1000)

def get_point_along_spline(dist):
    """Return the closest point on the spline a given a length along a spline."""
    idx = np.argmin(np.abs(spline.point_data['arc_length'] - dist))
    return spline.points[idx]

# distances along the spline we're interested in
dists = [0, 4, 8, 11]

# make labels
labels = []
points = []
for dist in dists:
    point = get_point_along_spline(dist)
    labels.append(f'Dist {dist}: ({point[0]:.2f}, {point[1]:.2f}, {point[2]:.2f})')
    points.append(point)

# plot it!
pl = pv.Plotter()
pl.add_mesh(spline, scalars='arc_length', render_lines_as_tubes=True, line_width=10)
pl.add_point_labels(points, labels, always_visible=True, point_size=20, render_points_as_spheres=True)
pl.show_bounds()
pl.show_axes()
pl.camera_position='xz'
pl.show()
```

**Reply by AlexanderJuestel** [reply]
That seems like a good approximation. I could increase the number of points used to create the spline in the first place to a resolution of approx. 10 cm, which is more than sufficient for us. I will try to apply your workflow to the geological logs as well and mark the discussion as answered then


Name of formations and coordinates had to be removed :) 
![image](https://user-images.githubusercontent.com/45469915/182812431-a8700986-1694-49b1-8212-31c5f4ece8b7.png)

**Reply by akaszynski** [reply]
Thanks for posting your solution here!

**Reply by MatthewFlamm** [reply]
Nice! To get the radius of the tube to depend on your data like you mentioned, you can explicitly use the tube filter, which accepts a `scalars` argument.  The `radius` arg sets the min radius in your data and the data is scaled relative to the minimum of the scalars.  There is also a `radius_factor` which can adjust the maximum radius.  It took some fiddling to do what I wanted.

**Comment by andrea-bistacchi** [comment]
Hello, this is incredibly useful, thanks!

Now we'd like to make one step further and be able to visualise **logs as ribbons**, either to the right or to the left of the well, filled or not, with a homogeneous colour or a colour scale mapping a property as in the tube log above. This is useful because with ribbons you can show multiple parameters at the same time.

We are trying with ribbon objects, but we were wondering if there is a way to keep the ribbon always **oriented facing the camera**, otherwise if it is fixed with respect to the XYZ axes, at some camera rotation angles it will be invisible or very thin.

The problem I see is that a vtkRibbon is a vtkPolydata in the end, but objects that can be kept oriented towards the camera with vtkFollower are vtkActors, correct?

Thanks!

**Comment by AlexanderJuestel** [comment]
Hello @akaszynski,

I was in the process of writing a follow-up question to this here until I realized that I had solved my issue myself. I will post a solution to the original problem in a little bit as I managed to make it working now :) 

![image](https://github.com/pyvista/pyvista/assets/45469915/4130339a-c199-4013-8f33-872e22d85fc1)
Image of a well constrained by three points (in red), resampled in between the red points and then assigned measurement (random) values (radii) that could represent measurements along the well based on a  slightly modified version of your `get_point_along_spline` function
